File: src-tauri/src/ipc/update.rs
//! 更新检查的 IPC 命令（D23）。
//!
//! 人类侧在设置页手动检查更新、查看版本、忽略某版本、配置更新源。
//! 自动检查在应用启动时由 `run()` 触发（后台、静默失败）。

use std::sync::Arc;

use tauri::State;

use crate::state::AppState;
use crate::update;

use super::IpcResult;

/// 更新检查的当前配置与状态（供设置页展示）。
#[derive(Debug, serde::Serialize)]
pub struct UpdateInfo {
    /// 当前应用版本。
    pub current_version: String,
    /// 更新源地址（可能为空，表示未配置）。
    pub source_url: String,
    /// 是否启用启动时自动检查。
    pub auto_check: bool,
    /// 用户忽略的版本。
    pub ignored_version: Option<String>,
    /// 上次检查结果（若有缓存）。
    pub last_result: Option<update::UpdateStatus>,
}

async fn info_inner(state: &AppState) -> crate::error::Result<UpdateInfo> {
    let conn = state.db.lock().await;
    Ok(UpdateInfo {
        current_version: update::current_version(),
        source_url: update::source_url(&conn)?,
        auto_check: update::auto_check_enabled(&conn)?,
        ignored_version: update::ignored_version(&conn)?,
        last_result: update::cached_result(&conn)?,
    })
}

/// 读取更新检查的配置与上次结果。
#[tauri::command]
pub async fn update_info(state: State<'_, Arc<AppState>>) -> Result<IpcResult<UpdateInfo>, ()> {
    Ok(match info_inner(&state).await {
        Ok(v) => IpcResult::ok(v),
        Err(e) => IpcResult::from(e),
    })
}

/// 检查更新。
///
/// `force` 为真表示用户手动触发：忽略 24 小时缓存，且失败时把原因告知用户；
/// 为假则复用缓存、失败静默（供启动时自动检查使用）。
#[tauri::command]
pub async fn update_check(
    force: bool,
    state: State<'_, Arc<AppState>>,
) -> Result<IpcResult<update::UpdateStatus>, ()> {
    // 分阶段加锁：先在锁内读取配置，再在锁外做网络请求，最后写回缓存。
    // 避免网络请求（最长 5 秒）期间占用数据库锁。
    let (url, ignored) = {
        let conn = state.db.lock().await;
        match (update::source_url(&conn), update::ignored_version(&conn)) {
            (Ok(u), Ok(i)) => (u, i),
            (Err(e), _) | (_, Err(e)) => return Ok(IpcResult::from(e)),
        }
    };

    // 缓存命中直接返回（仅自动检查走这条路）。
    if !force {
        let conn = state.db.lock().await;
        if update::cache_is_fresh(&conn).unwrap_or(false) {
            if let Ok(Some(cached)) = update::cached_result(&conn) {
                return Ok(IpcResult::ok(cached));
            }
        }
    }

    let status = match update::fetch_manifest(&url).await {
        Ok(manifest) => {
            let current = update::current_version();
            update::evaluate(&manifest, &current, ignored.as_deref(), update::platform_key())
        }
        Err(e) => {
            // 自动检查失败静默，仅记录；手动检查把原因返回给用户。
            if !force {
                tracing::debug!("自动更新检查失败（已忽略）：{e}");
            }
            return Ok(IpcResult::ok(update::UpdateStatus::Failed {
                reason: e.to_string(),
            }));
        }
    };

    // 写回缓存（短暂加锁）。
    {
        let conn = state.db.lock().await;
        if let Err(e) = crate::store::db::set_setting(
            &conn,
            update::SETTING_LAST_CHECK,
            &chrono::Utc::now().to_rfc3339(),
        ) {
            tracing::warn!("写入更新检查时间失败：{e}");
        }
        if let Ok(json) = serde_json::to_string(&status) {
            let _ = crate::store::db::set_setting(&conn, update::SETTING_CACHED, &json);
        }
    }

    Ok(IpcResult::ok(status))
}

/// 忽略某个版本（同版本不再提示）。
#[tauri::command]
pub async fn update_ignore_version(
    version: String,
    state: State<'_, Arc<AppState>>,
) -> Result<IpcResult<bool>, ()> {
    let conn = state.db.lock().await;
    Ok(match update::ignore_version(&conn, &version) {
        Ok(()) => IpcResult::ok(true),
        Err(e) => IpcResult::from(e),
    })
}

/// 设置更新源地址。
#[tauri::command]
pub async fn update_set_source(
    url: String,
    state: State<'_, Arc<AppState>>,
) -> Result<IpcResult<bool>, ()> {
    let conn = state.db.lock().await;
    Ok(match update::set_source_url(&conn, &url) {
        Ok(()) => IpcResult::ok(true),
        Err(e) => IpcResult::from(e),
    })
}

/// 设置是否启用启动时自动检查。
#[tauri::command]
pub async fn update_set_auto_check(
    enabled: bool,
    state: State<'_, Arc<AppState>>,
) -> Result<IpcResult<bool>, ()> {
    let conn = state.db.lock().await;
    Ok(match update::set_auto_check(&conn, enabled) {
        Ok(()) => IpcResult::ok(true),
        Err(e) => IpcResult::from(e),
    })
}

File: src/stores/update.ts
import { defineStore } from "pinia";
import { ref, computed } from "vue";

import { call } from "@/lib/ipc";
import { useAppStore } from "./app";

/**
 * 更新检查状态（D23）。
 *
 * 设计要点：
 * - 不做自动下载/安装，只提示并提供下载地址（客户决定）
 * - 手动检查会把失败原因展示给用户；自动检查静默
 * - 支持「忽略此版本」，同版本不再提示
 */
export interface UpdateInfo {
  current_version: string;
  /** **生效的**更新源地址（内置默认或用户自定义）。 */
  source_url: string;
  /** 是否为用户自定义；false 表示正在使用内置默认地址（D42）。 */
  source_is_custom: boolean;
  auto_check: boolean;
  ignored_version: string | null;
  last_result: UpdateStatus | null;
}

export type UpdateStatus =
  | { status: "up_to_date"; current: string; latest: string }
  | {
      status: "available";
      current: string;
      latest: string;
      notes: string | null;
      published_at: string | null;
      /** 该版本的 Release 页面地址（D58）；null 表示清单既没给也推导不出。 */
      release_url: string | null;
      sha256: string | null;
      size: number | null;
    }
  | { status: "ignored"; current: string; latest: string }
  | { status: "failed"; reason: string };

export const useUpdateStore = defineStore("update", () => {
  const info = ref<UpdateInfo | null>(null);
  const result = ref<UpdateStatus | null>(null);
  const checking = ref(false);
  /** 手动检查失败时的提示（自动检查不显示）。 */
  const manualError = ref<string | null>(null);

  const app = useAppStore();

  const currentVersion = computed(() => info.value?.current_version ?? "—");
  const hasUpdate = computed(() => result.value?.status === "available");

  async function refresh() {
    try {
      info.value = await call<UpdateInfo>("update_info");
      // 已有缓存结果时直接展示，避免设置页空白。
      if (info.value?.last_result && !result.value) {
        result.value = info.value.last_result;
      }
    } catch (e) {
      app.fail(e);
    }
  }

  /** 检查更新。`manual` 为真时展示失败原因。 */
  async function check(manual = true) {
    checking.value = true;
    manualError.value = null;
    try {
      const r = await call<UpdateStatus>("update_check", { force: manual });
      // "已忽略"不作为当前提示展示，保持界面清爽。
      result.value = r;
      if (manual && r.status === "available") {
        app.notify(`发现新版本 ${r.latest}`);
      }
      if (manual && r.status === "up_to_date") {
        app.notify("已是最新版本");
      }
      if (manual && r.status === "failed") {
        manualError.value = r.reason;
      }
    } catch (e) {
      if (manual) manualError.value = e instanceof Error ? e.message : String(e);
    } finally {
      checking.value = false;
    }
  }

  async function ignoreVersion(version: string) {
    try {
      await call<boolean>("update_ignore_version", { version });
      await refresh();
      // 忽略后立即清除提示，避免仍然显示。
      result.value = { status: "ignored", current: currentVersion.value, latest: version };
      app.notify(`已忽略版本 ${version}`);
    } catch (e) {
      app.fail(e);
    }
  }

  async function setSource(url: string) {
    try {
      await call<boolean>("update_set_source", { url });
      await refresh();
      app.notify("更新源已保存");
    } catch (e) {
      app.fail(e);
      throw e;
    }
  }

  /** 恢复内置默认更新源（清空自定义值即为回退，见 D42）。 */
  async function resetSource() {
    try {
      await call<boolean>("update_set_source", { url: "" });
      await refresh();
      app.notify("已恢复内置默认更新源");
    } catch (e) {
      app.fail(e);
      throw e;
    }
  }

  async function setAutoCheck(enabled: boolean) {
    try {
      await call<boolean>("update_set_auto_check", { enabled });
      await refresh();
    } catch (e) {
      app.fail(e);
      throw e;
    }
  }

  return {
    info,
    result,
    checking,
    manualError,
    currentVersion,
    hasUpdate,
    refresh,
    check,
    ignoreVersion,
    setSource,
    resetSource,
    setAutoCheck,
  };
});
